### PR TITLE
blockdev_mirror_hotunplug.py: change the critical condition on check device deleted.

### PR DESCRIPTION
**Pull request by nickzhq**
blockdev_mirror_hotunplug: change the critical condition on check device deleted.

ID: 2035988

Signed-off-by: Houqi (Nick) Zuo hzuo@redhat.com

### CODE REVIEW

**Comment by nickzhq** [comment]
Hello @zhencliu  ,please take a look, thanks!

**Comment by zhencliu** [comment]
hi @nickzhq , 
1. commit header is too long and also please check if line of commit msg is too long
2. I remembered we talked about this case before, some comments from my side, ping me offline if you have any concern

**Comment by nickzhq** [comment]
@zhencliu  please check it out, thanks!

File: qemu/tests/blockdev_mirror_hotunplug.py [hunk 1]
```diff
@@ -6,9 +6,12 @@
...
 
 class BlockdevMirrorHotunplugTest(BlockdevMirrorNowaitTest):
-    """
-    Block mirror with hotunplug test
-    """
+    def __init__(self, test, params, env):
+        """
+        Block mirror with hotunplug test
+        """
+        super().__init__(test, params, env)
```
Comment (zhencliu) [inline qemu/tests/blockdev_mirror_hotunplug.py:13]: super(BlockdevMirrorHotunplugTest, self).__init__

File: qemu/tests/blockdev_mirror_hotunplug.py [hunk 2]
```diff
@@ -20,13 +23,22 @@ def _device_del(device):
...
-            for item in self.main_vm.monitor.query("block"):
-                if device in item["qdev"]:
-                    return False
-            return True
+            res = self.main_vm.monitor.query("block")
+            device_numbers = len(res)
+            if device_numbers + 1 == self._device_numbers:
+                for item in res:
+                    if device not in item["qdev"]:
+                        return True
```
Comment (zhencliu) [inline qemu/tests/blockdev_mirror_hotunplug.py:40]: we cannot return True by checking one and only one item, loop all.

**Comment by nickzhq** [comment]
@zhencliu  , the code has been changed.   
Please check it out, thanks!

**Comment by nickzhq** [comment]
(1/1) Host_RHEL.m8.u7.qcow2.virtio_blk.up.virtio_net.Guest.RHEL.9.0.0.ppc64le.io-github-autotest-qemu.blockdev_mirror_hotunplug: STARTED
/ (1/1) Host_RHEL.m8.u7.qcow2.virtio_blk.up.virtio_net.Guest.RHEL.9.0.0.ppc64le.io-github-autotest-qemu.blockdev_mirror_hotunplug: PASS (317.08 s)
RESULTS    : PASS 1 | ERROR 0 | FAIL 0 | SKIP 0 | WARN 0 | INTERRUPT 0 | CANCEL 0

**Comment by nickzhq** [comment]
Hello @zhencliu   , the "return value" has been updated!

**Comment by zhencliu** [comment]
hi @aliang123 @MiriamDeng , would you please help to review this patch? Thanks.

**Comment by zhencliu** [comment]
LGTM

@nickzhq please make the commit header shorter :)

**Comment by MiriamDeng** [comment]
Test passed on ppc
- (1/1) Host_RHEL.m8.u6.product_av.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.9.0.0.ppc64le.io-github-autotest-qemu.blockdev_mirror_hotunplug: PASS (341.40 s)

**Comment by ngu-niny** [comment]
LGTM. Test PASS:

(1/1) Host_RHEL.m9.u0.nographic.qcow2.virtio_blk.up.virtio_net.Guest.RHEL.9.0.0.s390x.io-github-autotest-qemu.blockdev_mirror_hotunplug.s390-virtio: PASS (90.05 s)

File: qemu/tests/blockdev_mirror_hotunplug.py [hunk 3]
```diff
@@ -20,13 +22,23 @@ def _device_del(device):
...
+        self._device_numbers = len(self.main_vm.monitor.query("block"))
+
     def wait_till_frontend_devices_deleted(self):
         """Wait till devices removed from output of query-block"""
         def _is_device_deleted(device):
-            for item in self.main_vm.monitor.query("block"):
-                if device in item["qdev"]:
-                    return False
-            return True
+            res = self.main_vm.monitor.query("block")
```
Comment (aliang123) [inline qemu/tests/blockdev_mirror_hotunplug.py:34]: I think,
  if device in item.get("qdev"):
        return False
is enough.

After check with ngu, the issue was caused as ppc can't get complete info in a immediate state for the deleted device. 
So if we can't get qdev, the device has been deleted in fact.
Or another way, if we can't get qdev, we can break this loop, enter next _is_device_deleted.

Comment (zhencliu) [reply qemu/tests/blockdev_mirror_hotunplug.py:34]: > I think, if device in item.get("qdev"): return False is enough.
> 
> After check with ngu, the issue was caused as ppc can't get complete info in a immediate state for the deleted device. So if we can't get qdev, the device has been deleted in fact. 
>>> It's in an unstable state, maybe removing the device is still in progress.
Or another way, if we can't get qdev, we can break this loop, enter next _is_device_deleted.
>>> Safe way! Skip the unstable state, wait till the device is deleted. This is also OK from my side.

Comment (zhencliu) [reply qemu/tests/blockdev_mirror_hotunplug.py:34]: hi @nickzhq , what do you think about Aihua's comments?  It looks a simple solution.

**Comment by aliang123** [comment]
(1/1) Host_RHEL.m9.u1.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.9.0.0.x86_64.io-github-autotest-qemu.blockdev_mirror_hotunplug.q35: PASS (211.92 s)

Acked-by: Aihua Liang <aliang@redhat.com>

@ngu-niny @MiriamDeng, please help to check in your platform, thanks.

File: qemu/tests/blockdev_mirror_hotunplug.py [hunk 4]
```diff
@@ -1,15 +1,10 @@
 import time
 
 from virttest.qemu_monitor import QMPCmdError
-
```
Comment (zhencliu) [inline qemu/tests/blockdev_mirror_hotunplug.py:4]: pls keep the current code style, you can submit another commit if you have concern about the code style

File: qemu/tests/blockdev_mirror_hotunplug.py [hunk 5]
```diff
@@ -1,15 +1,10 @@
...
 from provider.blockdev_mirror_nowait import BlockdevMirrorNowaitTest
 
 
 class BlockdevMirrorHotunplugTest(BlockdevMirrorNowaitTest):
-    """
-    Block mirror with hotunplug test
-    """
-
```
Comment (zhencliu) [inline qemu/tests/blockdev_mirror_hotunplug.py:12]: The same comment as above, and it's better to make a brief introduction to the class

File: qemu/tests/blockdev_mirror_hotunplug.py [hunk 6]
```diff
@@ -24,7 +19,7 @@ def wait_till_frontend_devices_deleted(self):
         """Wait till devices removed from output of query-block"""
         def _is_device_deleted(device):
             for item in self.main_vm.monitor.query("block"):
-                if device in item["qdev"]:
+                if device in item.get("qdev"):
```
Comment (zhencliu) [inline qemu/tests/blockdev_mirror_hotunplug.py:22]: item.get("qdev") can return None, then we may hit TypeError exception with this code change.

The simplest way is to change item.get("qdev") to item.get("qdev", device), but it's hard to understand, so we should make it clear with comments that qdev can be absent while the device hotunplug is still in progress, in this situation, we will wait till all block info is removed

**Comment by zhencliu** [comment]
LGTM 

hi @ngu-niny , would you please help to verify this tc again on the host where we hit the 'qdev gone' issue? Thanks. 
hi @aliang123 , pls also make a double check, thanks.

**Comment by aliang123** [comment]
ack

**Comment by ngu-niny** [comment]
Test PASS:
 
(2/2) Host_RHEL.m8.u7.nographic.qcow2.virtio_blk.up.virtio_net.Guest.RHEL.9.0.0.s390x.io-github-autotest-qemu.blockdev_mirror_hotunplug.s390-virtio:  PASS (99.97 s)
